src/creatures/players/player.cpp

```diff
@@ -6571,6 +6579,42 @@ void Player::setSkullTicks(int64_t ticks) {
 	skullTicks = ticks;
 }
 
+Player::SkullTimeInfo Player::computeSkullTimeFromLastKill() const {
+	SkullTimeInfo info;
+	int64_t ticks = skullTicks;
+
+	if (ticks == 0 && (getSkull() == SKULL_RED || getSkull() == SKULL_BLACK)) {
+		const auto query = fmt::format(
+			"SELECT `time` FROM `player_kills` WHERE `player_id` = {} ORDER BY `time` DESC LIMIT 1;",
+			guid
+		);
+		if (auto result = g_database().storeQuery(query); result && result->hasNext()) {
```

Comment:
I have one final review to conduct on direct database access. I will thoroughly analyze our resolution and proceed with a comprehensive new review.